sql: TestListProfilerExecutionDetails failed

sql.TestListProfilerExecutionDetails [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/11863840?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/11863840?buildTab=artifacts#/) on master @ [30bc4f9afcd23a4c6aeba0fd592541f3b82f30b8](https://github.com/cockroachdb/cockroach/commits/30bc4f9afcd23a4c6aeba0fd592541f3b82f30b8):


```
=== RUN   TestListProfilerExecutionDetails
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/4e9b6c662b8a17b1b0bf70dc9d006949/logTestListProfilerExecutionDetails3920191949
    test_log_scope.go:81: use -show-logs to present logs inline
    conditional_wrap.go:190: 
        pkg/sql_test/pkg/sql/jobs_profiler_execution_details_test.go:328: (TestListProfilerExecutionDetails)
        	NOTICE: .ExecutorConfig() called via implicit interface ApplicationLayerInterface;
        HINT: consider using .ApplicationLayer().ExecutorConfig() instead.
    conditional_wrap.go:194: TIP: consider replacing the test server initialization from:
            ts, ... := serverutils.StartServer(t, ...)
            defer ts.Stopper().Stop(...)
        to:
            srv, ... := serverutils.StartServer(t, ...)
            defer srv.Stopper().Stop(...)
            ts := srv.ApplicationLayer()
        
        See also: https://go.crdb.dev/p/testserver-and-cluster-virtualization
--- FAIL: TestListProfilerExecutionDetails (4.13s)
=== RUN   TestListProfilerExecutionDetails/list_execution_detail_files
    jobs_profiler_execution_details_test.go:382: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/sql_test/pkg/sql/jobs_profiler_execution_details_test.go:382
        	Error:      	"[1/resumer-trace/20230921_010716.78~cockroach.sql.jobs.jobspb.TraceData.binpb 1/resumer-trace/20230921_010716.78~cockroach.sql.jobs.jobspb.TraceData.binpb.txt distsql.20230921_010716.55.html distsql.20230921_010716.89.html goroutines.20230921_010716.57.txt goroutines.20230921_010716.90.txt trace.20230921_010716.58.zip trace.20230921_010716.92.zip]" should have 10 item(s), but has 8
        	Test:       	TestListProfilerExecutionDetails/list_execution_detail_files
=== NAME  TestListProfilerExecutionDetails
    jobs_profiler_execution_details_test.go:394: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/4e9b6c662b8a17b1b0bf70dc9d006949/logTestListProfilerExecutionDetails3920191949
    --- FAIL: TestListProfilerExecutionDetails/list_execution_detail_files (0.54s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestListProfilerExecutionDetails.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-31713

sql.TestListProfilerExecutionDetails [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/11868060?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/11868060?buildTab=artifacts#/) on master @ [02435d6896d1ed53a47aeee62e4687aab8aed239](https://github.com/cockroachdb/cockroach/commits/02435d6896d1ed53a47aeee62e4687aab8aed239):


```
=== RUN   TestListProfilerExecutionDetails
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/4e9b6c662b8a17b1b0bf70dc9d006949/logTestListProfilerExecutionDetails2473475591
    test_log_scope.go:81: use -show-logs to present logs inline
    conditional_wrap.go:190: 
        pkg/sql_test/pkg/sql/jobs_profiler_execution_details_test.go:328: (TestListProfilerExecutionDetails)
        	NOTICE: .ExecutorConfig() called via implicit interface ApplicationLayerInterface;
        HINT: consider using .ApplicationLayer().ExecutorConfig() instead.
    conditional_wrap.go:194: TIP: consider replacing the test server initialization from:
            ts, ... := serverutils.StartServer(t, ...)
            defer ts.Stopper().Stop(...)
        to:
            srv, ... := serverutils.StartServer(t, ...)
            defer srv.Stopper().Stop(...)
            ts := srv.ApplicationLayer()
        
        See also: https://go.crdb.dev/p/testserver-and-cluster-virtualization
--- FAIL: TestListProfilerExecutionDetails (2.63s)
=== RUN   TestListProfilerExecutionDetails/list_execution_detail_files
    jobs_profiler_execution_details_test.go:382: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/sql_test/pkg/sql/jobs_profiler_execution_details_test.go:382
        	Error:      	"[1/resumer-trace/20230921_112105.24~cockroach.sql.jobs.jobspb.TraceData.binpb 1/resumer-trace/20230921_112105.24~cockroach.sql.jobs.jobspb.TraceData.binpb.txt distsql.20230921_112104.95.html distsql.20230921_112105.38.html goroutines.20230921_112104.96.txt goroutines.20230921_112105.38.txt trace.20230921_112104.97.zip trace.20230921_112105.39.zip]" should have 10 item(s), but has 8
        	Test:       	TestListProfilerExecutionDetails/list_execution_detail_files
=== NAME  TestListProfilerExecutionDetails
    jobs_profiler_execution_details_test.go:394: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/4e9b6c662b8a17b1b0bf70dc9d006949/logTestListProfilerExecutionDetails2473475591
    --- FAIL: TestListProfilerExecutionDetails/list_execution_detail_files (0.58s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestListProfilerExecutionDetails.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql.TestListProfilerExecutionDetails [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11918975?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11918975?buildTab=artifacts#/) on master @ [cb7c66f651909c0a0850cf8e89cdefc91e22a005](https://github.com/cockroachdb/cockroach/commits/cb7c66f651909c0a0850cf8e89cdefc91e22a005):


```
=== RUN   TestListProfilerExecutionDetails
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/b287de7194e17a0386ee1775f0d198ba/logTestListProfilerExecutionDetails3953966529
    test_log_scope.go:81: use -show-logs to present logs inline
    conditional_wrap.go:190: 
        pkg/sql_test/pkg/sql/jobs_profiler_execution_details_test.go:328: (TestListProfilerExecutionDetails)
        	NOTICE: .ExecutorConfig() called via implicit interface ApplicationLayerInterface;
        HINT: consider using .ApplicationLayer().ExecutorConfig() instead.
    conditional_wrap.go:194: TIP: consider replacing the test server initialization from:
            ts, ... := serverutils.StartServer(t, ...)
            defer ts.Stopper().Stop(...)
        to:
            srv, ... := serverutils.StartServer(t, ...)
            defer srv.Stopper().Stop(...)
            ts := srv.ApplicationLayer()
        
        See also: https://go.crdb.dev/p/testserver-and-cluster-virtualization
--- FAIL: TestListProfilerExecutionDetails (1.73s)
=== RUN   TestListProfilerExecutionDetails/list_execution_detail_files
    jobs_profiler_execution_details_test.go:382: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/sql_test/pkg/sql/jobs_profiler_execution_details_test.go:382
        	Error:      	"[1/resumer-trace/20230925_131619.30~cockroach.sql.jobs.jobspb.TraceData.binpb 1/resumer-trace/20230925_131619.30~cockroach.sql.jobs.jobspb.TraceData.binpb.txt distsql.20230925_131619.05.html distsql.20230925_131619.41.html goroutines.20230925_131619.06.txt goroutines.20230925_131619.43.txt trace.20230925_131619.08.zip trace.20230925_131619.45.zip]" should have 10 item(s), but has 8
        	Test:       	TestListProfilerExecutionDetails/list_execution_detail_files
=== NAME  TestListProfilerExecutionDetails
    jobs_profiler_execution_details_test.go:394: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/b287de7194e17a0386ee1775f0d198ba/logTestListProfilerExecutionDetails3953966529
    --- FAIL: TestListProfilerExecutionDetails/list_execution_detail_files (0.55s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestListProfilerExecutionDetails.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql.TestListProfilerExecutionDetails [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/11919306?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/11919306?buildTab=artifacts#/) on master @ [654a38dcf80184421b1c4a567391ed79085feb61](https://github.com/cockroachdb/cockroach/commits/654a38dcf80184421b1c4a567391ed79085feb61):


```
=== RUN   TestListProfilerExecutionDetails
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/4e9b6c662b8a17b1b0bf70dc9d006949/logTestListProfilerExecutionDetails4233885013
    test_log_scope.go:81: use -show-logs to present logs inline
    conditional_wrap.go:190: 
        pkg/sql_test/pkg/sql/jobs_profiler_execution_details_test.go:328: (TestListProfilerExecutionDetails)
        	NOTICE: .ExecutorConfig() called via implicit interface ApplicationLayerInterface;
        HINT: consider using .ApplicationLayer().ExecutorConfig() instead.
    conditional_wrap.go:194: TIP: consider replacing the test server initialization from:
            ts, ... := serverutils.StartServer(t, ...)
            defer ts.Stopper().Stop(...)
        to:
            srv, ... := serverutils.StartServer(t, ...)
            defer srv.Stopper().Stop(...)
            ts := srv.ApplicationLayer()
        
        See also: https://go.crdb.dev/p/testserver-and-cluster-virtualization
--- FAIL: TestListProfilerExecutionDetails (2.03s)
=== RUN   TestListProfilerExecutionDetails/list_execution_detail_files
    jobs_profiler_execution_details_test.go:382: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/sql_test/pkg/sql/jobs_profiler_execution_details_test.go:382
        	Error:      	"[1/resumer-trace/20230925_135618.41~cockroach.sql.jobs.jobspb.TraceData.binpb 1/resumer-trace/20230925_135618.41~cockroach.sql.jobs.jobspb.TraceData.binpb.txt distsql.20230925_135618.06.html distsql.20230925_135618.54.html goroutines.20230925_135618.09.txt goroutines.20230925_135618.60.txt trace.20230925_135618.12.zip trace.20230925_135618.69.zip]" should have 10 item(s), but has 8
        	Test:       	TestListProfilerExecutionDetails/list_execution_detail_files
=== NAME  TestListProfilerExecutionDetails
    jobs_profiler_execution_details_test.go:394: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/4e9b6c662b8a17b1b0bf70dc9d006949/logTestListProfilerExecutionDetails4233885013
    --- FAIL: TestListProfilerExecutionDetails/list_execution_detail_files (0.89s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestListProfilerExecutionDetails.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/jobs_profiler_execution_details_test.go
+++ b/pkg/sql/jobs_profiler_execution_details_test.go
@@ -198,7 +198,8 @@ func TestReadWriteProfilerExecutionDetails(t *testing.T) {
 		runner.QueryRow(t, `IMPORT INTO t CSV DATA ('nodelocal://1/foo') WITH DETACHED`).Scan(&importJobID)
 		<-isRunning
 		runner.Exec(t, `SELECT crdb_internal.request_job_execution_details($1)`, importJobID)
-		distSQLDiagram := checkExecutionDetails(t, s, jobspb.JobID(importJobID), "distsql")
+		distSQLDiagram, err := checkExecutionDetails(t, s, jobspb.JobID(importJobID), "distsql")
+		require.NoError(t, err)
 		require.Regexp(t, "<meta http-equiv=\"Refresh\" content=\"0\\; url=https://cockroachdb\\.github\\.io/distsqlplan/decode.html.*>", string(distSQLDiagram))
 		close(continueRunning)
 		jobutils.WaitForJobToSucceed(t, runner, jobspb.JobID(importJobID))
@@ -224,7 +225,8 @@ func TestReadWriteProfilerExecutionDetails(t *testing.T) {
 		runner.QueryRow(t, `IMPORT INTO t CSV DATA ('nodelocal://1/foo') WITH DETACHED`).Scan(&importJobID)
 		<-blockCh
 		runner.Exec(t, `SELECT crdb_internal.request_job_execution_details($1)`, importJobID)
-		goroutines := checkExecutionDetails(t, s, jobspb.JobID(importJobID), "goroutines")
+		goroutines, err := checkExecutionDetails(t, s, jobspb.JobID(importJobID), "goroutines")
+		require.NoError(t, err)
 		continueCh <- struct{}{}
 		jobutils.WaitForJobToSucceed(t, runner, jobspb.JobID(importJobID))
 		require.True(t, strings.Contains(string(goroutines), fmt.Sprintf("labels: {\"foo\":\"bar\", \"job\":\"IMPORT id=%d\", \"n\":\"1\"}", importJobID)))
@@ -249,7 +251,14 @@ func TestReadWriteProfilerExecutionDetails(t *testing.T) {
 		var importJobID int
 		runner.QueryRow(t, `IMPORT INTO t CSV DATA ('nodelocal://1/foo') WITH DETACHED`).Scan(&importJobID)
 		jobutils.WaitForJobToSucceed(t, runner, jobspb.JobID(importJobID))
-		trace := checkExecutionDetails(t, s, jobspb.JobID(importJobID), "resumer-trace")
+		var trace []byte
+		// There may be some delay between the job finishing and the trace being
+		// persisted.
+		testutils.SucceedsSoon(t, func() error {
+			var err error
+			trace, err = checkExecutionDetails(t, s, jobspb.JobID(importJobID), "resumer-trace")
+			return err
+		})
 		require.Contains(t, string(trace), "should see this")
 	})
 
@@ -273,7 +282,8 @@ func TestReadWriteProfilerExecutionDetails(t *testing.T) {
 		runner.QueryRow(t, `IMPORT INTO t CSV DATA ('nodelocal://1/foo') WITH DETACHED`).Scan(&importJobID)
 		<-blockCh
 		runner.Exec(t, `SELECT crdb_internal.request_job_execution_details($1)`, importJobID)
-		activeTraces := checkExecutionDetails(t, s, jobspb.JobID(importJobID), "trace")
+		activeTraces, err := checkExecutionDetails(t, s, jobspb.JobID(importJobID), "trace")
+		require.NoError(t, err)
 		continueCh <- struct{}{}
 		jobutils.WaitForJobToSucceed(t, runner, jobspb.JobID(importJobID))
 		unzip, err := zip.NewReader(bytes.NewReader(activeTraces), int64(len(activeTraces)))
@@ -369,6 +379,14 @@ func TestListProfilerExecutionDetails(t *testing.T) {
 		continueCh <- struct{}{}
 		jobutils.WaitForJobToPause(t, runner, jobspb.JobID(importJobID))
 
+		testutils.SucceedsSoon(t, func() error {
+			files = listExecutionDetails(t, s, jobspb.JobID(importJobID))
+			if len(files) != 5 {
+				return errors.Newf("expected 5 files, got %d: %v", len(files), files)
+			}
+			return nil
+		})
+
 		// Resume the job, so it can write another DistSQL diagram and goroutine
 		// snapshot.
 		runner.Exec(t, `SET CLUSTER SETTING jobs.debug.pausepoints = ''`)
@@ -378,8 +396,13 @@ func TestListProfilerExecutionDetails(t *testing.T) {
 		runner.Exec(t, `SELECT crdb_internal.request_job_execution_details($1)`, importJobID)
 		continueCh <- struct{}{}
 		jobutils.WaitForJobToSucceed(t, runner, jobspb.JobID(importJobID))
-		files = listExecutionDetails(t, s, jobspb.JobID(importJobID))
-		require.Len(t, files, 10)
+		testutils.SucceedsSoon(t, func() error {
+			files = listExecutionDetails(t, s, jobspb.JobID(importJobID))
+			if len(files) != 10 {
+				return errors.Newf("expected 10 files, got %d: %v", len(files), files)
+			}
+			return nil
+		})
 		require.Regexp(t, "[0-9]/resumer-trace/.*~cockroach\\.sql\\.jobs\\.jobspb\\.TraceData\\.binpb", files[0])
 		require.Regexp(t, "[0-9]/resumer-trace/.*~cockroach\\.sql\\.jobs\\.jobspb\\.TraceData\\.binpb.txt", files[1])
 		require.Regexp(t, "[0-9]/resumer-trace/.*~cockroach\\.sql\\.jobs\\.jobspb\\.TraceData\\.binpb", files[2])
@@ -423,30 +446,44 @@ func listExecutionDetails(
 
 func checkExecutionDetails(
 	t *testing.T, s serverutils.TestServerInterface, jobID jobspb.JobID, filename string,
-) []byte {
+) ([]byte, error) {
 	t.Helper()
 
 	client, err := s.GetAdminHTTPClient()
-	require.NoError(t, err)
+	if err != nil {
+		return nil, err
+	}
 
 	url := s.AdminURL().String() + fmt.Sprintf("/_status/job_profiler_execution_details/%d?%s", jobID, filename)
 	req, err := http.NewRequest("GET", url, nil)
-	require.NoError(t, err)
+	if err != nil {
+		return nil, err
+	}
 
 	req.Header.Set("Content-Type", httputil.ProtoContentType)
 	resp, err := client.Do(req)
-	require.NoError(t, err)
+	if err != nil {
+		return nil, err
+	}
 	defer resp.Body.Close()
 	body, err := io.ReadAll(resp.Body)
-	require.NoError(t, err)
+	if err != nil {
+		return nil, err
+	}
 	require.Equal(t, http.StatusOK, resp.StatusCode)
 
 	edResp := serverpb.GetJobProfilerExecutionDetailResponse{}
-	require.NoError(t, protoutil.Unmarshal(body, &edResp))
+	if err := protoutil.Unmarshal(body, &edResp); err != nil {
+		return nil, err
+	}
 
 	r := bytes.NewReader(edResp.Data)
 	data, err := io.ReadAll(r)
-	require.NoError(t, err)
-	require.NotEmpty(t, data)
-	return data
+	if err != nil {
+		return data, err
+	}
+	if len(data) == 0 {
+		return data, errors.New("no data returned")
+	}
+	return data, nil
 }

